File: backend/app/schemas/procurement.py
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel, field_validator


# ─── Purchase Order ───────────────────────────────────────────────────────────

class PurchaseOrderItemCreate(BaseModel):
    paper_material_id: int | None = None
    other_material_id: int | None = None
    ten_hang: str | None = None
    so_cuon: int | None = None
    so_luong: Decimal
    dvt: str | None = None
    don_gia: Decimal = Decimal(0)
    ghi_chu: str | None = None

    @field_validator("so_luong")
    @classmethod
    def sl_duong(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("Số lượng phải > 0")
        return v


class PurchaseOrderItemUpdate(BaseModel):
    ten_hang: str | None = None
    so_cuon: int | None = None
    so_luong: Decimal | None = None
    dvt: str | None = None
    don_gia: Decimal | None = None
    ghi_chu: str | None = None


class PurchaseOrderItemResponse(BaseModel):
    id: int
    order_id: int
    paper_material_id: int | None = None
    other_material_id: int | None = None
    ten_hang: str | None = None
    # Joined
    ten_nguyen_lieu: str | None = None
    ma_nguyen_lieu: str | None = None
    so_cuon: int | None = None
    so_luong: Decimal
    dvt: str | None = None
    don_gia: Decimal
    thanh_tien: Decimal
    so_luong_da_nhap: Decimal
    ghi_chu: str | None = None

    class Config:
        from_attributes = True


class PurchaseOrderCreate(BaseModel):
    loai_don: str  # giay_cuon | khac
    ngay_dat: date
    supplier_id: int
    nv_thu_mua_id: int | None = None
    ten_nhom_hang: str | None = None
    noi_dung: str | None = None
    ghi_chu: str | None = None
    items: list[PurchaseOrderItemCreate] = []


class PurchaseOrderUpdate(BaseModel):
    ngay_dat: date | None = None
    supplier_id: int | None = None
    nv_thu_mua_id: int | None = None
    ten_nhom_hang: str | None = None
    noi_dung: str | None = None
    ghi_chu: str | None = None


class PurchaseOrderListItem(BaseModel):
    id: int
    so_don_mua: str
    loai_don: str
    ngay_dat: date
    supplier_id: int
    ten_nha_cung_cap: str | None = None
    tong_tien: Decimal
    trang_thai: str
    so_dong: int
    created_at: datetime

    class Config:
        from_attributes = True


class PurchaseOrderResponse(BaseModel):
    id: int
    so_don_mua: str
    loai_don: str
    ngay_dat: date
    supplier_id: int
    ten_nha_cung_cap: str | None = None
    nv_thu_mua_id: int | None = None
    ten_nv_thu_mua: str | None = None
    nguoi_duyet_id: int | None = None
    ten_nguoi_duyet: str | None = None
    ngay_duyet: datetime | None = None
    ten_nhom_hang: str | None = None
    tong_tien: Decimal
    trang_thai: str
    noi_dung: str | None = None
    ghi_chu: str | None = None
    items: list[PurchaseOrderItemResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class POPagedResponse(BaseModel):
    items: list[PurchaseOrderListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


# ─── Material Receipt ─────────────────────────────────────────────────────────

class MaterialReceiptItemCreate(BaseModel):
    purchase_order_item_id: int | None = None
    paper_material_id: int | None = None
    other_material_id: int | None = None
    ten_hang: str | None = None
    so_luong: Decimal
    dvt: str | None = None
    don_gia: Decimal = Decimal(0)
    ghi_chu: str | None = None

    @field_validator("so_luong")
    @classmethod
    def sl_duong(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("Số lượng phải > 0")
        return v


class MaterialReceiptItemUpdate(BaseModel):
    ten_hang: str | None = None
    so_luong: Decimal | None = None
    dvt: str | None = None
    don_gia: Decimal | None = None
    ghi_chu: str | None = None


class MaterialReceiptItemResponse(BaseModel):
    id: int
    receipt_id: int
    purchase_order_item_id: int | None = None
    paper_material_id: int | None = None
    other_material_id: int | None = None
    ten_hang: str | None = None
    ten_nguyen_lieu: str | None = None
    ma_nguyen_lieu: str | None = None
    so_luong: Decimal
    dvt: str | None = None
    don_gia: Decimal
    thanh_tien: Decimal
    ghi_chu: str | None = None

    class Config:
        from_attributes = True


class MaterialReceiptCreate(BaseModel):
    ngay_nhap: date
    phan_xuong: str | None = None
    warehouse_id: int
    supplier_id: int
    purchase_order_id: int | None = None
    so_phieu_can: str | None = None
    bien_so_xe: str | None = None
    trong_luong_xe: Decimal | None = None
    trong_luong_hang: Decimal | None = None
    ghi_chu: str | None = None
    items: list[MaterialReceiptItemCreate] = []


class MaterialReceiptUpdate(BaseModel):
    ngay_nhap: date | None = None
    phan_xuong: str | None = None
    warehouse_id: int | None = None
    supplier_id: int | None = None
    so_phieu_can: str | None = None
    bien_so_xe: str | None = None
    trong_luong_xe: Decimal | None = None
    trong_luong_hang: Decimal | None = None
    ghi_chu: str | None = None


class MaterialReceiptListItem(BaseModel):
    id: int
    so_phieu: str
    ngay_nhap: date
    supplier_id: int
    ten_nha_cung_cap: str | None = None
    ten_kho: str | None = None
    purchase_order_id: int | None = None
    so_don_mua: str | None = None
    tong_tien: Decimal
    trang_thai: str
    so_dong: int
    created_at: datetime

    class Config:
        from_attributes = True


class MaterialReceiptResponse(BaseModel):
    id: int
    so_phieu: str
    ngay_nhap: date
    phan_xuong: str | None = None
    warehouse_id: int
    ten_kho: str | None = None
    supplier_id: int
    ten_nha_cung_cap: str | None = None
    purchase_order_id: int | None = None
    so_don_mua: str | None = None
    so_phieu_can: str | None = None
    bien_so_xe: str | None = None
    trong_luong_xe: Decimal | None = None
    trong_luong_hang: Decimal | None = None
    tong_tien: Decimal
    ghi_chu: str | None = None
    trang_thai: str
    items: list[MaterialReceiptItemResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ReceiptPagedResponse(BaseModel):
    items: list[MaterialReceiptListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


# ─── Inventory Summary ────────────────────────────────────────────────────────

class MaterialInventoryRow(BaseModel):
    ma_nguyen_lieu: str
    ten_nguyen_lieu: str
    loai: str  # giay_cuon | khac
    dvt: str | None = None
    ton_luong: Decimal
    gia_tri_ton: Decimal
    don_gia_binh_quan: Decimal

File: frontend/src/api/procurement.ts
import client from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface POItemCreate {
  paper_material_id?: number | null
  other_material_id?: number | null
  ten_hang?: string | null
  so_cuon?: number | null
  so_luong: number
  dvt?: string | null
  don_gia?: number
  ghi_chu?: string | null
}

export interface POItemResponse {
  id: number
  order_id: number
  paper_material_id: number | null
  other_material_id: number | null
  ten_hang: string | null
  ten_nguyen_lieu: string | null
  ma_nguyen_lieu: string | null
  so_cuon: number | null
  so_luong: number
  dvt: string | null
  don_gia: number
  thanh_tien: number
  so_luong_da_nhap: number
  ghi_chu: string | null
}

export interface POCreate {
  loai_don: string  // giay_cuon | khac
  ngay_dat: string
  supplier_id: number
  nv_thu_mua_id?: number | null
  ten_nhom_hang?: string | null
  noi_dung?: string | null
  ghi_chu?: string | null
  items: POItemCreate[]
}

export interface POUpdate {
  ngay_dat?: string
  supplier_id?: number
  nv_thu_mua_id?: number | null
  ten_nhom_hang?: string | null
  noi_dung?: string | null
  ghi_chu?: string | null
}

export interface POListItem {
  id: number
  so_don_mua: string
  loai_don: string
  ngay_dat: string
  supplier_id: number
  ten_nha_cung_cap: string | null
  tong_tien: number
  trang_thai: string
  so_dong: number
  created_at: string
}

export interface POResponse {
  id: number
  so_don_mua: string
  loai_don: string
  ngay_dat: string
  supplier_id: number
  ten_nha_cung_cap: string | null
  nv_thu_mua_id: number | null
  ten_nv_thu_mua: string | null
  nguoi_duyet_id: number | null
  ten_nguoi_duyet: string | null
  ngay_duyet: string | null
  ten_nhom_hang: string | null
  tong_tien: number
  trang_thai: string
  noi_dung: string | null
  ghi_chu: string | null
  items: POItemResponse[]
  created_at: string
  updated_at: string
}

export interface POPagedResponse {
  items: POListItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// ─── Material Receipt Types ───────────────────────────────────────────────────

export interface ReceiptItemCreate {
  purchase_order_item_id?: number | null
  paper_material_id?: number | null
  other_material_id?: number | null
  ten_hang?: string | null
  so_luong: number
  dvt?: string | null
  don_gia?: number
  ghi_chu?: string | null
}

export interface ReceiptItemResponse {
  id: number
  receipt_id: number
  purchase_order_item_id: number | null
  paper_material_id: number | null
  other_material_id: number | null
  ten_hang: string | null
  ten_nguyen_lieu: string | null
  ma_nguyen_lieu: string | null
  so_luong: number
  dvt: string | null
  don_gia: number
  thanh_tien: number
  ghi_chu: string | null
}

export interface ReceiptCreate {
  ngay_nhap: string
  phan_xuong?: string | null
  warehouse_id: number
  supplier_id: number
  purchase_order_id?: number | null
  so_phieu_can?: string | null
  bien_so_xe?: string | null
  trong_luong_xe?: number | null
  trong_luong_hang?: number | null
  ghi_chu?: string | null
  items: ReceiptItemCreate[]
}

export interface ReceiptListItem {
  id: number
  so_phieu: string
  ngay_nhap: string
  supplier_id: number
  ten_nha_cung_cap: string | null
  ten_kho: string | null
  purchase_order_id: number | null
  so_don_mua: string | null
  tong_tien: number
  trang_thai: string
  so_dong: number
  created_at: string
}

export interface ReceiptResponse {
  id: number
  so_phieu: string
  ngay_nhap: string
  phan_xuong: string | null
  warehouse_id: number
  ten_kho: string | null
  supplier_id: number
  ten_nha_cung_cap: string | null
  purchase_order_id: number | null
  so_don_mua: string | null
  so_phieu_can: string | null
  bien_so_xe: string | null
  trong_luong_xe: number | null
  trong_luong_hang: number | null
  tong_tien: number
  ghi_chu: string | null
  trang_thai: string
  items: ReceiptItemResponse[]
  created_at: string
  updated_at: string
}

export interface ReceiptPagedResponse {
  items: ReceiptListItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface MaterialInventoryRow {
  ma_nguyen_lieu: string
  ten_nguyen_lieu: string
  loai: string
  dvt: string | null
  ton_luong: number
  gia_tri_ton: number
  don_gia_binh_quan: number
}

// ─── API ──────────────────────────────────────────────────────────────────────

export const procurementApi = {
  // Purchase Orders
  listPO: (params?: {
    search?: string
    loai_don?: string
    trang_thai?: string
    supplier_id?: number
    tu_ngay?: string
    den_ngay?: string
    page?: number
    page_size?: number
  }) => client.get<POPagedResponse>('/procurement/purchase-orders', { params }),

  getPO: (id: number) =>
    client.get<POResponse>(`/procurement/purchase-orders/${id}`),

  createPO: (data: POCreate) =>
    client.post<POResponse>('/procurement/purchase-orders', data),

  updatePO: (id: number, data: POUpdate) =>
    client.put<POResponse>(`/procurement/purchase-orders/${id}`, data),

  approvePO: (id: number) =>
    client.patch<POResponse>(`/procurement/purchase-orders/${id}/approve`),

  cancelPO: (id: number) =>
    client.patch(`/procurement/purchase-orders/${id}/cancel`),

  // Material Receipts
  listReceipts: (params?: {
    search?: string
    trang_thai?: string
    supplier_id?: number
    warehouse_id?: number
    tu_ngay?: string
    den_ngay?: string
    page?: number
    page_size?: number
  }) => client.get<ReceiptPagedResponse>('/procurement/material-receipts', { params }),

  getReceipt: (id: number) =>
    client.get<ReceiptResponse>(`/procurement/material-receipts/${id}`),

  createReceipt: (data: ReceiptCreate) =>
    client.post<ReceiptResponse>('/procurement/material-receipts', data),

  confirmReceipt: (id: number) =>
    client.patch<ReceiptResponse>(`/procurement/material-receipts/${id}/confirm`),

  // Inventory
  getMaterialInventory: (params?: {
    warehouse_id?: number
    loai?: string
    search?: string
  }) => client.get<MaterialInventoryRow[]>('/procurement/inventory/material', { params }),
}

// ─── Constants ────────────────────────────────────────────────────────────────

export const PO_TRANG_THAI: Record<string, { label: string; color: string }> = {
  cho_duyet:  { label: 'Chờ duyệt',    color: 'orange' },
  da_duyet:   { label: 'Đã duyệt',     color: 'blue' },
  hoan_thanh: { label: 'Hoàn thành',   color: 'green' },
  huy:        { label: 'Đã hủy',       color: 'red' },
}

export const PO_LOAI: Record<string, string> = {
  giay_cuon: 'Giấy cuộn',
  khac:      'Hàng khác',
}

export const RECEIPT_TRANG_THAI: Record<string, { label: string; color: string }> = {
  nhap:     { label: 'Nháp',          color: 'default' },
  xac_nhan: { label: 'Đã xác nhận',  color: 'green' },
}
